lib/xml.js

```diff
@@ -93,103 +94,79 @@ var escapeUserFields = function(field, value) {
-    return null;
-  }
-  if (common.isString(zObject)) {
-    return zObject;
-  }
-  var result = '';
-  var znstype = '';
-  var element = '';
-  var ns = '';
-
-  if (type.toLowerCase() === 'request') {
-    element = 'requests';
-    ns = ns1;
-  } else {
-    element = 'zObjects';
-    ns = ns1;
-    znstype = 'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="' + ns2 + ':' + type + '"';
-  }
-
-  var fields = xmlFromFieldList(zObject, ns2);
-
-  result = tag(element, ns, fields, znstype);
-  return result;
-};
+function convert (method, object, objectType) {
+  var zObjectMethods = ['create', 'update', 'generate'];
+  var ns1Actions = ['delete', 'execute'];
 
+  var encapsulate = function(key, obj) {
+    var res = {};
+    res[key] = obj;
+    return res;
+  };
 
-/**
- * make XML for deletes
- *
- * in: [{id: 123}, {id: 456}]
- * out: <zns:delete><zns:type>Contact</zns:type><zns:ids>123</zns:ids><zns:ids>456</zns:ids></zns:delete>
- */
-var makeDeletePayload = function (zObjectType, zObjects) {
-  if (common.isString(zObjects)) return zObjects;
-  zObjects.unshift({type: zObjectType}); // TODO: clone array to avoid modifying original
-  var swapIds = function(zObj) {
-    if (!zObj.ids && !zObj.Ids) {
-      if (zObj.id) { return {ids: zObj.id}; }
-      if (zObj.Id) { return {ids: zObj.Id}; }
-    }
-    return zObj;
+  // Main function to transform object into XML.
+  var toXml = function(obj, depth) {
+    return _.join(_.map(obj, function(value, key) {
+      var prefix, name, val, attributes;
+      // If value is an array, manipulate it to generate a correct XML
+      // Depth is not to be increased here due to the way we handle the array.
+      if (_.isArray(value)) {
+        return _.join(_.map(value, function(v) {
+          var o = {};
+          o[key] = v;
+          return toXml(o, depth);
+        }), '\n');
+      } else {
+        // If method is one of create, update or generate, handle zObject tags at 1st level
+        if (_.includes(zObjectMethods, method) && _.includes(_.map(objectList, function(obj) {return obj.name.toLowerCase();}), key.toLowerCase())) {
+          if (depth === 1) {
+            name = 'zObjects';
+            prefix = ns1;
+            attributes = 'xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="' + ns2 + ':' + key + '"';
+          } else {
+            name = key;
+            prefix = ns1;
+            attributes = 'xsi:type="' + ns2 + ':' + key + '"';
+          }
+        } else {
+          name = key;
+        }
+        // For actions delete and execute, all tags are using the ns1 prefix
+        if (_.includes(ns1Actions, method)) {
+          prefix = prefix || ns1;
+        }
+
+        // If value is an object, simply reapply toXml method to transform it to XML increasing depth
+        if (_.isObject(value)) {
+          prefix = prefix || ns1;
+          depth++;
+          val = toXml(value, depth);
```

Comment:
As `depth` is not used outside the recursive call of `toXml`
The two previous lines could be updated to:
`val = toXml(value, depth+1)` or `val = toXml(value, ++depth)`